File: models_impl/src/main/java/com/mdgd/pokemon/models/repo/PokemonsRepository.java
package com.mdgd.pokemon.models.repo;

import com.mdgd.pokemon.models.infra.Result;
import com.mdgd.pokemon.models.repo.cache.PokemonsCache;
import com.mdgd.pokemon.models.repo.dao.PokemonsDao;
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema;
import com.mdgd.pokemon.models.repo.network.Network;

import java.util.List;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public class PokemonsRepository implements PokemonsRepo {
    private final PokemonsDao dao;
    private final Network network;
    private final PokemonsCache cache;

    public PokemonsRepository(PokemonsDao pokemonsDao, Network pokemonsNetwork, PokemonsCache cache) {
        this.dao = pokemonsDao;
        this.network = pokemonsNetwork;
        this.cache = cache;
    }

    @Override
    public Single<Result<List<PokemonFullDataSchema>>> getPage(Integer page) {
        return dao.getPage(page, PAGE_SIZE)
                .flatMap(result -> {
                    if (result.isError()) {
                        return Single.just(new Result<>(result.getError()));
                    } else {
                        final List<PokemonFullDataSchema> list = result.getValue();
                        return list.isEmpty() ? loadPage(page) : Single.just(new Result<>(list));
                    }
                });
    }

    @Override
    public Observable<Result<Long>> loadPokemons() {
        final long count = dao.getCount();
        return network.getPokemonsCount()
                .flatMapObservable(result -> {
                    if (result.isError()) {
                        return Observable.just(new Result<>(result.getError()));
                    } else if (count == result.getValue()) {
                        return Observable.just(new Result<>(count));
                    } else {
                        return loadPokemonsInner();
                    }
                });

    }

    @Override
    public List<PokemonFullDataSchema> getPokemons() {
        return cache.getPokemons();
    }

    private Observable<Result<Long>> loadPokemonsInner() {
        return network.loadPokemons(PAGE_SIZE * 3)
                .flatMap(result -> {
                    if (result.isError()) {
                        return Observable.just(new Result<>(result.getError()));
                    } else {
                        return dao.save(result.getValue())
                                .toSingleDefault(new Result<>((long) result.getValue().size()))
                                .toObservable();
                    }
                });
    }

    private Single<Result<List<PokemonFullDataSchema>>> loadPage(Integer page) {
        return network.loadPokemons(page, PAGE_SIZE)
                .flatMap(result -> {
                    if (result.isError()) {
                        return Single.just(new Result<>(result.getError()));
                    } else {
                        return dao.save(result.getValue())
                                .andThen(dao.getPage(page, PAGE_SIZE))
                                .doOnEvent((e, error) -> {
                                    if (!result.isError()) {
                                        if (page == 0) {
                                            cache.setPokemons(e.getValue());
                                        } else {
                                            cache.addPokemons(e.getValue());
                                        }
                                    }
                                });
                    }
                });
    }
}

File: models_impl/src/main/java/com/mdgd/pokemon/models/repo/PokemonsRepository.kt
package com.mdgd.pokemon.models.repo

import com.mdgd.pokemon.models.infra.Result
import com.mdgd.pokemon.models.repo.dao.PokemonsDao
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema
import com.mdgd.pokemon.models.repo.network.Network
import com.mdgd.pokemon.models.repo.network.schemas.PokemonDetails
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableSource
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.core.SingleSource
import io.reactivex.rxjava3.functions.Function

class PokemonsRepository(private val dao: PokemonsDao, private val network: Network) : PokemonsRepo {

    override fun getPage(page: Int): Single<Result<List<PokemonFullDataSchema>>> {
        return dao.getPage(page, PokemonsRepo.PAGE_SIZE)
                .flatMap(Function<Result<List<PokemonFullDataSchema>>, SingleSource<Result<List<PokemonFullDataSchema>>>> { result: Result<List<PokemonFullDataSchema>> ->
                    if (result.isError()) {
                        return@Function Single.just(Result<List<PokemonFullDataSchema>>(result.getError()))
                    } else {
                        val list = result.getValue()
                        return@Function if (list.isEmpty()) loadPage(page) else Single.just(Result(list))
                    }
                })
    }

    override fun loadPokemons(): Observable<Result<Long>> {
        val count = dao.getCount()
        return network.getPokemonsCount()
                .flatMapObservable { result: Result<Long> ->
                    if (result.isError()) {
                        return@flatMapObservable Observable.just(Result<Long>(result.getError()))
                    } else if (count == result.getValue()) {
                        return@flatMapObservable Observable.just(Result(count))
                    } else {
                        return@flatMapObservable loadPokemonsInner()
                    }
                }
    }

    private fun loadPokemonsInner(): Observable<Result<Long>> {
        return network.loadPokemons(PokemonsRepo.PAGE_SIZE * 3)
                .flatMap(Function<Result<List<PokemonDetails>>, ObservableSource<Result<Long>>> { result: Result<List<PokemonDetails>> ->
                    if (result.isError()) {
                        return@Function Observable.just(Result<Long>(result.getError()))
                    } else {
                        return@Function dao.save(result.getValue())
                                .toSingleDefault(Result(result.getValue().size.toLong()))
                                .toObservable()
                    }
                })
    }

    private fun loadPage(page: Int): Single<Result<List<PokemonFullDataSchema>>> {
        return network.loadPokemons(page, PokemonsRepo.PAGE_SIZE)
                .flatMap(Function<Result<List<PokemonDetails>>, SingleSource<out Result<List<PokemonFullDataSchema>>>> { result: Result<List<PokemonDetails>> ->
                    if (result.isError()) {
                        return@Function Single.just(Result<List<PokemonFullDataSchema>>(result.getError()))
                    } else {
                        return@Function dao.save(result.getValue())
                                .andThen(dao.getPage(page, PokemonsRepo.PAGE_SIZE))
                    }
                })
    }
}
